executor/src/witgen/jit/affine_symbolic_expression.rs

```diff
@@ -0,0 +1,430 @@
+    pub fn solve(&self, range_constraints: &impl RangeConstraintSet<V, T>) -> Vec<Effect<T, V>> {
+        match self.coefficients.len() {
+            0 => {
+                return if self.offset.is_known_zero() {
+                    vec![]
+                } else {
+                    // TODO this is a non-satisfiable constraint - should we panic?
+                    vec![]
+                };
+            }
+            1 => {
+                let (var, coeff) = self.coefficients.iter().next().unwrap();
+                assert!(!coeff.is_known_zero());
+                let value = &self.offset / &coeff.neg();
+                return vec![Effect::Assignment(var.clone(), value)];
+            }
+            _ => {}
+        }
+
+        let r = self.solve_through_constraints(range_constraints);
+        if !r.is_empty() {
+            return r;
+        }
+        let negated = -self;
+        let r = negated.solve_through_constraints(range_constraints);
+        if !r.is_empty() {
+            return r;
+        }
+        self.transfer_constraints(range_constraints)
+            .into_iter()
+            .chain(self.transfer_constraints(range_constraints))
+            .collect()
+    }
+
+    /// Tries to solve a bit-decomposition equation.
+    fn solve_through_constraints(
+        &self,
+        range_constraints: &impl RangeConstraintSet<V, T>,
+    ) -> Vec<Effect<T, V>> {
+        let constrained_coefficients = self
+            .coefficients
+            .iter()
+            .filter_map(|(var, coeff)| {
+                coeff.try_to_number().and_then(|c| {
+                    range_constraints
+                        .range_constraint(var.clone())
+                        .map(|rc| (var.clone(), c, rc))
+                })
+            })
+            .collect_vec();
+
+        // All the coefficients need to have known range constraints.
+        if constrained_coefficients.len() != self.coefficients.len() {
+            return Vec::new();
+        }
+
+        // Check if they are mutually exclusive and compute assignments.
+        let mut covered_bits: <T as FieldElement>::Integer = 0.into();
+        let mut assignments = vec![];
+        for (var, coeff, constraint) in constrained_coefficients {
+            let mask = *constraint.multiple(coeff).mask();
+            if !(mask & covered_bits).is_zero() {
+                // Overlapping range constraints.
+                return vec![];
+            } else {
+                covered_bits |= mask;
+            }
+            let masked = &(-&self.offset) & &T::from(mask).into();
+            assignments.push(Effect::Assignment(
+                var.clone(),
+                masked.integer_div(&coeff.into()),
+            ));
+        }
+
+        if covered_bits >= T::modulus() {
+            return vec![];
+        }
+
+        // TODO we need to add an assertion that offset is covered by the masks.
```

Comment:
The problem is that `offset` is usually a run-time value. So we could do a run-time check and if the check fails, the system is conflicting. Or we just continue to solve and ignore the conflict...